# LSE Machine Learning: Practical Applications
## Module 5 Unit 1 IDE Activity (Practice)
### In this IDE activity, you have the opportunity to practise the execution of logistic regression on a data set in R.
As you complete this activity, you are required to read the text cells throughout the notebook and then run the code in the cells that follow. Be mindful of the syntax used to execute certain functionalities within R to produce a desired result. In completing this activity, you should gain the necessary practical skills to complete the IDE Activity (Assessment) that follows.

### Step 1: Load the relevant packages

The functionality to estimate logistic regression is in base R and does not have to be loaded separately. However, you need to load the tidyverse package for data manipulation purposes. To interpret the coefficients in the logistic regression, the `margins` package is used.

In [9]:
#!install.packages("margins") 

library(tidyverse)
library(margins)  #to obtain the average marginal effects

Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.0     ✔ purrr   0.3.4
✔ tibble  3.1.2     ✔ dplyr   1.0.4
✔ tidyr   1.1.3     ✔ stringr 1.4.0
✔ readr   1.4.0     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



### Step 2: Load the data

To execute logistic regression, load the credit card data set.

In [10]:
credit_data <- read.csv("Credit.csv")

Once the data set is loaded into R, use the `str()` function to analyse the structure of the data frame. Using the `head()` function, also consider the first few rows of the data.

In [11]:
str(credit_data)
head(credit_data)

'data.frame':	10000 obs. of  4 variables:
 $ default: Factor w/ 2 levels "No","Yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ student: Factor w/ 2 levels "No","Yes": 1 2 1 1 1 2 1 2 1 1 ...
 $ balance: num  730 817 1074 529 786 ...
 $ income : num  44362 12106 31767 35704 38463 ...


default student balance   income   
1 No      No       729.5265 44361.625
2 No      Yes      817.1804 12106.135
3 No      No      1073.5492 31767.139
4 No      No       529.2506 35704.494
5 No      No       785.6559 38463.496
6 No      Yes      919.5885  7491.559

The first rows indicate that all the columns are relevant. However, the first two variables are categorical in nature and are not saved as factor variables. These variables must be saved as factor variables using the `as.factor()` function. Once this has been done, confirm that the new structure of the data is correct.

In [12]:
credit_data$default <- as.factor(credit_data$default)
credit_data$student <- as.factor(credit_data$student)
str(credit_data)

'data.frame':	10000 obs. of  4 variables:
 $ default: Factor w/ 2 levels "No","Yes": 1 1 1 1 1 1 1 1 1 1 ...
 $ student: Factor w/ 2 levels "No","Yes": 1 2 1 1 1 2 1 2 1 1 ...
 $ balance: num  730 817 1074 529 786 ...
 $ income : num  44362 12106 31767 35704 38463 ...


Once the data is ready, the logistic regression model can be fitted onto the data set. In this example, the ***default*** variable is the variable to be predicted, and all other variables are predictors.

### Step 3: Interpret the output

Use the `glm()` function to estimate the logistic regression model.

In [18]:
logitReg <- glm(default ~., data = credit_data, family = binomial(link = logit))
summary(logitReg)

ERROR: Error in glm.control(iter = 1000): unused argument (iter = 1000)


In this summary, the significance of the variables is shown, indicating that the ***student*** and ***balance*** varaibles are significant. The output shows that these variables are all statistically significant. To determine the relationship between these variables, the `margins()` function may be used. In the output the average marginal effect (AME) is the average change in **probability** with a unit increase in the predictor.

In [17]:
summary(margins(logitReg))

factor     AME           SE           z          p             lower        
1 balance     1.232348e-04 4.851866e-06 25.3994633 2.556535e-142  1.137253e-04
2 income      6.516621e-08 1.765025e-07  0.3692085  7.119723e-01 -2.807722e-07
3 studentYes -1.326965e-02 4.659183e-03 -2.8480647  4.398599e-03 -2.240148e-02
  upper        
1  1.327443e-04
2  4.111047e-07
3 -4.137823e-03

### Step 4: Make the prediction

Now that the model has been fitted and interpreted, it can be used to make predictions on a hypothetical individual. Predict the expected probability of a student, with a balance of USD2,000 and an income of USD30,000, to default on their credit card repayments.

In [15]:
newData <- data.frame("Yes", 2000, 30000)
colnames(newData) <- c("student", "balance", "income")
predict(logitReg, newData, type = "response")

1 
0.5120459

The output of this cell shows that the likelihood of this hypothetical individual defaulting on their credit card repayments is 51.2%.